File: Python/add2ints/add2ints.py
import vertica_sdk

class add2ints(vertica_sdk.ScalarFunctionFactory):
    """Return the sum of two integer columns"""

    def __init__(self):
        pass
    def setup(self, server_interface, col_types):
        pass
    def processBlock(self, server_interface, arg_reader, res_writer):
        # Writes a string to the UDx log file.
        server_interface.log("Python UDx - Adding 2 ints!")
        while(True):
            # Example of error checking best practices.
            product_id = arg_reader.getInt(2)
            if product_id < 100:
                raise ValueError("Invalid Product ID")
            if arg_reader.isNull(0) or arg_reader.isNull(1):
                raise ValueError("I found a NULL!")
            else:
                first_int = arg_reader.getInt(0) # Input column
                second_int = arg_reader.getInt(1) # Input column
            res_writer.setInt(first_int + second_int) # Sum of input columns.
            res_writer.next() # Read the next row.
            if not arg_reader.next():
                # Stop processing when there are no more input rows.
                break
    def destroy(self, server_interface, col_types):
        pass

class add2ints_factory(vertica_sdk.ScalarFunctionFactory):
    
    def createScalarFunction(self, srv):
        return add2ints()

    def getPrototype(self, srv_interface, arg_types, return_type):
        arg_types.addInt()
        arg_types.addInt()
        arg_types.addInt()
        return_type.addInt()

    def getReturnType(self, srv_interface, arg_types, return_type):
        return_type.addInt()

File: Java-and-C++/ScalarFunctions/Add2Ints.cpp
/* Copyright (c) 2005 - 2016 Hewlett Packard Enterprise Development LP  -*- C++ -*-
 *
 * Description: Example User Defined Scalar Function: Add 2 ints
 *
 * Create Date: Apr 29, 2011
 */
#include "Vertica.h"

class exception;
using namespace Vertica;

/*
 * This is a simple function that adds two integers and returns the result
 */
class Add2Ints : public ScalarFunction
{
public:

    /*
     * This method processes a block of rows in a single invocation.
     *
     * The inputs are retrieved via argReader
     * The outputs are returned via resWriter
     */
    virtual void processBlock(ServerInterface &srvInterface,
                              BlockReader &argReader,
                              BlockWriter &resWriter)
    {
        try {
            // Basic error checking
            if (argReader.getNumCols() != 2)
                vt_report_error(0, "Function only accept 2 arguments, but %zu provided", 
                                argReader.getNumCols());

            // While we have inputs to process
            do {
                const vint a = argReader.getIntRef(0);
                const vint b = argReader.getIntRef(1);
                resWriter.setInt(a+b);
                resWriter.next();
            } while (argReader.next());
        } catch(std::exception& e) {
            // Standard exception. Quit.
            vt_report_error(0, "Exception while processing block: [%s]", e.what());
        }
    }
};

class Add2IntsFactory : public ScalarFunctionFactory
{
    // return an instance of Add2Ints to perform the actual addition.
    virtual ScalarFunction *createScalarFunction(ServerInterface &interface)
    { return vt_createFuncObject<Add2Ints>(interface.allocator); }

    virtual void getPrototype(ServerInterface &interface,
                              ColumnTypes &argTypes,
                              ColumnTypes &returnType)
    {
        argTypes.addInt();
        argTypes.addInt();
        returnType.addInt();
    }
};

RegisterFactory(Add2IntsFactory);
